inverse_covariance/profiling/graphs.py

```diff
@@ -0,0 +1,159 @@
+    row[1: 1 + degree] = sign_row * prng.uniform(low=low, high=high, size=degree)
+
+    # QUESTION to M:
+    # Why is this normalized this way (can sum between 1 and -1)?
+    # What if the row sums to 0 (unlikely, but could be)? 
+    row /= np.abs(np.sum(row))  
+
+    return sp.linalg.toeplitz(c=row, r=row)
+
+
+def blocks(prng, block, n_blocks=2, chain_blocks=True):
+    """Replicates `block` matrix n_blocks times diagonally to create a
+    square matrix of size n_features = block.size[0] * n_blocks and with zeros
+    along the diagonal.
+
+    The graph can be made fully connected using chaining assumption when 
+    chain_blocks=True (default).
+    
+    This utility can be used to generate cluster networks or banded lattice n
+    networks, among others.
+
+    Parameters
+    -----------        
+    block : 2D array (n_block_features, n_block_features) 
+        Prototype block adjacency matrix.
+
+    n_blocks : int (default=2)
+        Number of blocks.  Returned matrix will be square with 
+        shape n_block_features * n_blocks.
+
+    chain_blocks : bool (default=True)
+        Apply random lattice structure to chain blocks together.
+    """
+    n_block_features, _ = block.shape
+    n_features = n_block_features * n_blocks
+    adjacency = np.zeros((n_features, n_features))    
+        
+    dep_groups = np.eye(n_blocks)
+    if chain_blocks:
+        chain_alpha = np.round(0.01 + 0.5 / n_blocks, 2)
+        chain_groups = lattice(prng, n_blocks, chain_alpha, random_sign=False)
+        chain_groups *= -0.1
+        dep_groups += chain_groups
+        
+    adjacency = np.kron(dep_groups, block)
+    adjacency[np.where(np.eye(n_features))] = 0 
+    return adjacency
+
+
+def _to_diagonally_dominant(adjacency):
+    """Make unweighted adjacency matrix M diagonally dominant using the 
+    Laplacian.
+    """    
+    adjacency += np.diag(np.sum(adjacency != 0, axis=1) + 0.01)
+    return adjacency
+
+
+def _to_diagonally_dominant_weighted(adjacency):
+    """Make weighted adjacency matrix M diagonally dominant using the 
+    Laplacian.
+
+    QUESTION:  Which is it?
+    NOTE:  This was called make_diagonally_dominant() in examples
+           but weighted in matlab version.
+    """    
+    adjacency += np.diag(np.sum(np.abs(adjacency), axis=1) + 0.01)
+    return adjacency
+
+
+def _to_correlation(adjacency):
+    """Call only after diagonal dominance is ensured."""   
+    d = np.sqrt(np.diag(adjacency))
+    adjacency /= d
+    adjacency /= d[:, np.newaxis]
+    return adjacency
+
+
+class Graph(object):
+    """Base class that returns the adjacency matrix for a network via
+    the .create() method.
+
+    Parameters
+    ----------- 
+    seed : int
+        Seed for np.random.RandomState seed. (default=1)
+    """
+    def __init__(self, seed=1):
+        self.seed = seed
+        self.prng = np.random.RandomState(self.seed)
+
+    def to_precision(self, adjacency, weighted=True):
```

Comment:
yes, the `.to_covariance(self, precision)` was doing step 4.  Will rename and small refactor.